File: openmrs-client/src/main/java/org/openmrs/mobile/activities/patientdashboard/visits/PatientVisitsFragment.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */

package org.openmrs.mobile.activities.patientdashboard.visits;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.openmrs.android_sdk.library.OpenmrsAndroid;
import com.openmrs.android_sdk.library.models.Patient;
import com.openmrs.android_sdk.library.models.Visit;
import com.openmrs.android_sdk.utilities.ApplicationConstants;
import com.openmrs.android_sdk.utilities.ToastUtil;

import org.openmrs.mobile.R;
import org.openmrs.mobile.activities.patientdashboard.PatientDashboardActivity;
import org.openmrs.mobile.activities.patientdashboard.PatientDashboardContract;
import org.openmrs.mobile.activities.patientdashboard.PatientDashboardFragment;
import org.openmrs.mobile.activities.visitdashboard.VisitDashboardActivity;
import org.openmrs.mobile.databinding.FragmentPatientVisitBinding;

import java.util.List;

public class PatientVisitsFragment extends PatientDashboardFragment implements PatientDashboardContract.ViewPatientVisits {
    private RecyclerView visitRecyclerView;
    private TextView emptyList;
    private FragmentPatientVisitBinding binding =null;
    private PatientDashboardActivity mPatientDashboardActivity;
    public static final int REQUEST_CODE_FOR_VISIT = 1;
    private Patient patient;

    public static PatientVisitsFragment newInstance() {
        return new PatientVisitsFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        super.setPresenter(mPresenter);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.patients_visit_tab_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.actionStartVisit:
                if (OpenmrsAndroid.getSyncState()) {
                    ((PatientDashboardVisitsPresenter) mPresenter).syncVisits();
                } else {
                    ToastUtil.notify(getString(R.string.offline_mode_not_supported));
                }
                break;
            default:
                // Do nothing
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showErrorToast(String message) {
        ToastUtil.error(message);
    }

    @Override
    public void showErrorToast(int messageId) {
        ToastUtil.error(getString(messageId));
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentPatientVisitBinding.inflate(inflater, null, false);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        visitRecyclerView = binding.patientVisitRecyclerView;
        visitRecyclerView.setHasFixedSize(true);
        visitRecyclerView.setLayoutManager(linearLayoutManager);

        emptyList = binding.emptyVisitsList;
        return binding.getRoot();

    }

    public void startVisit() {
        if (patient.isDeceased()) {
            ToastUtil.error(getString(R.string.cannot_start_visit_for_deceased));
        } else {
            ((PatientDashboardVisitsPresenter) mPresenter).startVisit();
        }
    }

    @Override
    public void dismissCurrentDialog() {
        ((PatientDashboardActivity) getActivity()).dismissCustomFragmentDialog();
    }

    @Override
    public void toggleRecyclerListVisibility(boolean isVisible) {
        if (isVisible) {
            visitRecyclerView.setVisibility(View.VISIBLE);
            emptyList.setVisibility(View.GONE);
        } else {
            visitRecyclerView.setVisibility(View.GONE);
            emptyList.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void setVisitsToDisplay(List<Visit> visits) {
        PatientVisitsRecyclerViewAdapter adapter = new PatientVisitsRecyclerViewAdapter(this, visits);
        visitRecyclerView.setAdapter(adapter);
        visitRecyclerView.getAdapter().notifyDataSetChanged();
    }

    @Override
    public void goToVisitDashboard(Long visitID) {
        Intent intent = new Intent(this.getActivity(), VisitDashboardActivity.class);
        intent.putExtra(ApplicationConstants.BundleKeys.VISIT_ID, visitID);
        startActivityForResult(intent, REQUEST_CODE_FOR_VISIT);
    }

    @Override
    public void showStartVisitDialog(boolean isVisitPossible) {
        PatientDashboardActivity activity = (PatientDashboardActivity) getActivity();
        if (activity.getSupportActionBar() != null) {
            if (isVisitPossible) {
                activity.showStartVisitDialog(activity.getSupportActionBar().getTitle());
            } else {
                activity.showStartVisitImpossibleDialog(activity.getSupportActionBar().getTitle());
            }
        }
    }

    @Override
    public void showStartVisitProgressDialog() {
        ((PatientDashboardActivity) getActivity()).showProgressDialog(R.string.action_starting_visit);
    }

    @Override
    public void setPatient(Patient mPatient) {
        patient = mPatient;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            try {
                mPatientDashboardActivity.hideFABs(true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/patientdashboard/visits/PatientVisitsFragment.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package org.openmrs.mobile.activities.patientdashboard.visits

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.openmrs.android_sdk.library.models.OperationType.PatientVisitStarting
import com.openmrs.android_sdk.library.models.OperationType.PatientVisitsFetching
import com.openmrs.android_sdk.library.models.Result
import com.openmrs.android_sdk.library.models.Visit
import com.openmrs.android_sdk.utilities.ApplicationConstants
import com.openmrs.android_sdk.utilities.ApplicationConstants.BundleKeys.PATIENT_ID_BUNDLE
import com.openmrs.android_sdk.utilities.NetworkUtils.isOnline
import com.openmrs.android_sdk.utilities.ToastUtil.error
import com.openmrs.android_sdk.utilities.ToastUtil.notify
import dagger.hilt.android.AndroidEntryPoint
import org.openmrs.mobile.R
import org.openmrs.mobile.activities.BaseFragment
import org.openmrs.mobile.activities.patientdashboard.PatientDashboardActivity
import org.openmrs.mobile.activities.visitdashboard.VisitDashboardActivity
import org.openmrs.mobile.databinding.FragmentPatientVisitBinding
import org.openmrs.mobile.utilities.makeGone
import org.openmrs.mobile.utilities.makeVisible
import org.openmrs.mobile.utilities.observeOnce

@AndroidEntryPoint
class PatientVisitsFragment : BaseFragment() {
    private var _binding: FragmentPatientVisitBinding? = null
    private val binding get() = _binding!!

    private val viewModel: PatientDashboardVisitsViewModel by viewModels()

    private lateinit var patientDashboardActivity: PatientDashboardActivity

    override fun onAttach(context: Context) {
        super.onAttach(context)
        patientDashboardActivity = context as PatientDashboardActivity
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentPatientVisitBinding.inflate(inflater, null, false)

        setupAdapter()
        setupObserver()


        return binding.root
    }

    override fun onResume() {
        super.onResume()
        fetchPatientVisits()
    }

    private fun setupAdapter() {
        with(binding.patientVisitRecyclerView) {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(activity)
            adapter = PatientVisitsRecyclerViewAdapter(this@PatientVisitsFragment, emptyList())
        }
    }

    private fun setupObserver() {
        viewModel.result.observe(viewLifecycleOwner, Observer { result ->
            when (result) {
                is Result.Loading -> {
                    when (result.operationType) {
                        PatientVisitStarting -> showStartVisitProgressDialog()
                        else -> {
                        }
                    }
                }
                is Result.Success -> {
                    dismissCurrentDialog()
                    when (result.operationType) {
                        PatientVisitsFetching -> showVisitsList(result.data)
                        PatientVisitStarting -> goToVisitDashboard(result.data[0].id!!)
                        else -> {
                        }
                    }
                }
                is Result.Error -> {
                    dismissCurrentDialog()
                    when (result.operationType) {
                        PatientVisitsFetching -> showErrorFetchingVisits()
                        PatientVisitStarting -> error(getString(R.string.visit_start_error))
                        else -> {
                        }
                    }
                }
                else -> throw IllegalStateException()
            }
        })
    }

    private fun fetchPatientVisits() {
        viewModel.fetchVisitsData()
    }

    fun startVisit() {
        viewModel.startVisit()
    }

    private fun showVisitsList(visits: List<Visit>) {
        with(binding) {
            if (visits.isEmpty()) {
                patientVisitRecyclerView.makeGone()
                emptyVisitsList.makeVisible()
            } else {
                (binding.patientVisitRecyclerView.adapter as PatientVisitsRecyclerViewAdapter).updateList(visits)
                patientVisitRecyclerView.makeVisible()
                emptyVisitsList.makeGone()
            }
        }
    }

    private fun showErrorFetchingVisits() {
        with(binding) {
            patientVisitRecyclerView.makeGone()
            emptyVisitsList.makeVisible()
            emptyVisitsList.text = getString(R.string.get_patient_from_database_error)
            error(getString(R.string.get_patient_from_database_error))
        }
    }

    private fun showStartVisitStatus() {
        if (!isOnline()) notify(getString(R.string.offline_mode_not_supported))
        else if (viewModel.getPatient().isDeceased) {
            error(getString(R.string.cannot_start_visit_for_deceased))
        } else {
            viewModel.hasActiveVisit().observeOnce(viewLifecycleOwner, Observer { hasActiveVisit ->
                with(patientDashboardActivity.supportActionBar!!) {
                    if (hasActiveVisit) patientDashboardActivity.showStartVisitImpossibleDialog(title)
                    else patientDashboardActivity.showStartVisitDialog(title)
                }
            })
        }
    }

    private fun showStartVisitProgressDialog() {
        patientDashboardActivity.showProgressDialog(R.string.action_starting_visit)
    }

    private fun dismissCurrentDialog() {
        patientDashboardActivity.dismissCustomFragmentDialog()
    }

    fun goToVisitDashboard(visitID: Long) {
        Intent(activity, VisitDashboardActivity::class.java).apply {
            putExtra(ApplicationConstants.BundleKeys.VISIT_ID, visitID)
            startActivity(this)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.patients_visit_tab_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.actionStartVisit -> showStartVisitStatus()
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        fun newInstance(patientId: String): PatientVisitsFragment {
            val fragment = PatientVisitsFragment()
            fragment.arguments = bundleOf(Pair(PATIENT_ID_BUNDLE, patientId))
            return fragment
        }
    }
}
